### ISSUE TITLE
Neovim hangs when an open buffer is modified externally and foldexpr is enabled.

### ISSUE THREAD

**Issue by freddiehaddad** [neovim/neovim#25608]
### Problem

With treesitter installed and enabled, set the following options:

```lua
vim.opt.foldenable = false
vim.opt.foldmethod = 'expr'
vim.opt.foldexpr = 'v:lua.vim.treesitter.foldexpr()'
```

Install treesitter with the following options:

```lua
opts = {
	auto_install = true,
	highlight = {
		enable = true,
	},
	incremental_selection = {
		enable = true,
		keymaps = {
			init_selection = '<c-i>',
			node_incremental = '<c-i>',
			scope_incremental = false,
			node_decremental = '<bs>',
		},
	},
	indent = {
		enable = true,
	},
}
```

1. Modify a git-tracked file inside neovim and save it. Leave the buffer visible and open.
2. In a separate terminal, restore the file using git restore
3. Go back to neovim

It will hang with high cpu utilization.

### Steps to reproduce

With treesitter installed and enabled, set the following options:

```lua
vim.opt.foldenable = false
vim.opt.foldmethod = 'expr'
vim.opt.foldexpr = 'v:lua.vim.treesitter.foldexpr()'
```

Install treesitter with the following options:

```lua
opts = {
	auto_install = true,
	highlight = {
		enable = true,
	},
	incremental_selection = {
		enable = true,
		keymaps = {
			init_selection = '<c-i>',
			node_incremental = '<c-i>',
			scope_incremental = false,
			node_decremental = '<bs>',
		},
	},
	indent = {
		enable = true,
	},
}
```

1. Modify a git-tracked file inside neovim and save it. Leave the buffer visible and open.
2. In a separate terminal, restore the file using git restore
3. Go back to neovim

It will hang with high cpu utilization.

### Expected behavior

No hang.

### Neovim version (nvim -v)

0.9.4

### Vim (not Nvim) behaves the same?

NA

### Operating system/version

Arch Linux

### Terminal name/version

Alacritty / TMUX

### $TERM environment variable

tmux-256color

### Installation

system package manager

**Comment by clason** [comment]
Suffice it to say, I cannot reproduce this with just the three options and treesitter highlighting enabled.

**Comment by freddiehaddad** [comment]
I can do it with any file.. As far as a minimal config, it doesn't get more minimal than what I provided. I deleted all neovim files with:

```bash
rm -rfv "$HOME"/.local/{state,share}/nvim "$HOME"/.cache/nvim "$HOME"/.config/nvim/{lazy-lock,.luarc,package-lock}.json
```

Installed only nvim-treesitter and performed the steps I mentioned. It happens with ANY file. Edit and save it in neovim, leave neovim open and the buffer visible. Navigate to a separate terminal, and run a git restore on that file undoing any edits. Go back to the terminal window here the open buffer is up in neovim and it will hang.

A full config would be:

```lua
local lazypath = vim.fn.stdpath('data') .. '/lazy/lazy.nvim'
local lazy_is_installed = vim.loop.fs_stat(lazypath)
if not lazy_is_installed then vim.fn.system({
	'git',
	'clone',
	'--filter=blob:none',
	'https://github.com/folke/lazy.nvim.git',
	'--branch=stable',
	lazypath,
}) end
vim.opt.runtimepath:prepend(lazypath)

require('lazy').setup({
	{
		'nvim-treesitter/nvim-treesitter',
		build = ':TSUpdate',
		init = function()
			vim.opt.foldenable = false
			vim.opt.foldmethod = 'expr'
			vim.opt.foldexpr = 'v:lua.vim.treesitter.foldexpr()'
		end,
		main = 'nvim-treesitter.configs',
		opts = {
			auto_install = true,
			highlight = {
				enable = true,
			},
			incremental_selection = {
				enable = true,
				keymaps = {
					init_selection = '<c-i>',
					node_incremental = '<c-i>',
					scope_incremental = false,
					node_decremental = '<bs>',
				},
			},
			indent = {
				enable = false,
			},
		},
	},
})
```

1. Put this setup in an init.lua file being tracked by git and commit it.
2. Open this file in neovim and edit and save it.
3. Open a separate terminal and do a git restore
4. Go back to neovim, and it will hang.

100% reproducible.

FWIW, neovim eventually recovers, but there's a long period with high cpu and the editor completely unresponsive.

**Comment by justinmk** [comment]
Does it happen with `:set noautoread`? And if not, does `:edit` then trigger it? Then try `:edit!`.

**Comment by freddiehaddad** [comment]
With `noautoread`, when I return to neovim, I see the following prompt:

```
W11: Warning: File "lua/plugins/treesitter.lua" has changed since editing started
See ":help W11" for more info.
[O]K, (L)oad File, Load File (a)nd Options:
```

1. Upon selecting L, neovim hangs.
2. If I select O instead, neovim does NOT hang. Typing `:edit` after triggers the hang.
2.1. If I type `:edit!` instead, neovim hangs.

In all scenarios, the following settings NOT prevents the hang.

```lua
vim.opt.foldenable = false
vim.opt.foldmethod = 'expr'
vim.opt.foldexpr = 'v:lua.vim.treesitter.foldexpr()'
```

**Comment by itaranto** [comment]
> Suffice it to say, I cannot reproduce this with just the three options and treesitter highlighting enabled.

This issue was opened because https://github.com/neovim/neovim/issues/24186 was closed. The minimal example to reproduce it remains the same: https://github.com/neovim/neovim/issues/24186#issuecomment-1612109306

(At least for me)

I'm sorry I couldn't debug (bisect) this any further, I'll try doing it during the weekend.

**Comment by freddiehaddad** [comment]
> Using 'nvim_treesitter#foldexpr()' instead of 'v:lua.vim.treesitter.foldexpr()' doesn't trigger the issue.
Can also confirm this.

**Comment by clason** [comment]
> This issue was opened because https://github.com/neovim/neovim/issues/24186 was closed.

That is why you should not open duplicate posts without clear indication. Can you reproduce it completely without nvim-treesitter (and lazy)?

> Using 'nvim_treesitter#foldexpr()' instead of 'v:lua.vim.treesitter.foldexpr()' doesn't trigger the issue.

That _is_ useful information. Can you bisect Neovim (you only need to check [these commits](https://github.com/neovim/neovim/commits/master/runtime/lua/vim/treesitter/_fold.lua), i.e. pass `runtime/lua/vim/treesitter/_fold.lua` to `git bisect`)?

**Comment by justinmk** [comment]
> With `noautoread`, when I return to neovim, I see the following prompt: `W11: Warning: File "lua/plugins/treesitter.lua" has changed ...`

Interesting, I wouldn't have expected that with 'autoread' _disabled_. Is `vim.treesitter.foldexpr()` triggering a file read? When `:edit` reloads the file, and `vim.treesitter.foldexpr()` also reloads the file, causing a loop?

> If I type `:edit!` instead, neovim also hangs.

Hmm, that suggests the `W11` prompt isn't part of the repro.

**Comment by itaranto** [comment]
> That is why you should not open duplicate posts without clear indication. Can you reproduce it completely without nvim-treesitter (and lazy)?

I didn't. 

In the previous issue I said that I couldn't reproduce it in `master` (at that time). 

But then I tested in 0.9.2 when it came out and I added a comment saying that.

**Comment by itaranto** [comment]
> That is useful information. Can you bisect Neovim (you only need to check [these commits](https://github.com/neovim/neovim/commits/master/runtime/lua/vim/treesitter/_fold.lua), i.e. pass runtime/lua/vim/treesitter/_fold.lua to git bisect)?

Yes, I'll do it when I can (tomorrow probably).

**Comment by justinmk** [comment]
> No, the file is read either automatically (with autoread) or manually (after entering `L`)

Then what is triggering the W11 prompt for the 'noautoread' case?

**Comment by freddiehaddad** [comment]
> (Again, this is why a minimal example is needed -- can you reproduce this with only foldexpr? Nvim-treesitter should not be part of these steps.)

Are you asking for a repro with the following inside `init.lua`:

```lua
vim.opt.foldenable = false
vim.opt.foldmethod = 'expr'
-- https://github.com/neovim/neovim/issues/25608
-- vim.opt.foldexpr = 'v:lua.vim.treesitter.foldexpr()'
vim.opt.foldexpr = 'nvim_treesitter#foldexpr()'
--

**Comment by clason** [comment]
@itaranto 

> The minimal.lua is the one I linked in my comment above, it doesn't use any plugin.

And does it _need_ to be the `go` parser? @freddiehaddad reports the issue with a Lua file.

(It's really difficult to handle two people reporting slightly different issues at the same time. I appreciate you trying to help, but right now we're talking about _this_ issue, with a related by different example.)

@freddiehaddad 

> Are you asking for a repro with the following inside init.lua:

I am asking for a _full_ `init.lua` I can run with `nvim --clean -u init.lua` and observe the problem you describe. The three lines by themselves are not enough to reproduce this (I tested).

**Comment by freddiehaddad** [comment]
With treesitter disabled, and the following vim settings set the issue does not occur:

``lua
vim.opt.foldenable = false
vim.opt.foldmethod = 'expr'
vim.opt.foldexpr = 'nvim_treesitter#foldexpr()'
```

With treesitter enabled, all settings disabled and foldexpr set to `'v:lua.vim.treesitter.foldexpr()'`, the issue happens:

```lua
opts = {
	auto_install = true,
	highlight = {
		enable = false,
	},
	incremental_selection = {
		enable = false,
		keymaps = {
			init_selection = '<c-i>',
			node_incremental = '<c-i>',
			scope_incremental = false,
			node_decremental = '<bs>',
		},
	},
	indent = {
		enable = false,
	},
},
```

However, in every case, setting foldexpr to `'nvim_treesitter#foldexpr()'` doesn't cause the hang.

**I was able to reproduce this by opening a file and undoing a change `u` from the undo file history.**

**Comment by clason** [comment]
Again, don't show snippets! Please provide a full, minimal config _I_ can test!

**Comment by freddiehaddad** [comment]
You have the config.. I just changed enabled to disabled.. do you really need a new config each time? Happy to provide that, but seems unnecessary.

**Comment by clason** [comment]
I need a _full_ config that can be tested with `nvim --clean -u test.lua`. You have not provided one in this issue so far.

**Comment by freddiehaddad** [comment]
> I need a full config that can be tested with nvim --clean -u test.lua. You have not provided one in this issue so far.

https://github.com/neovim/neovim/issues/25608#issuecomment-1759426086

**Comment by clason** [comment]
That is not minimal.

**Comment by itaranto** [comment]
> @itaranto
> 
> > The minimal.lua is the one I linked in my comment above, it doesn't use any plugin.
> 
> And does it _need_ to be the `go` parser? @freddiehaddad reports the issue with a Lua file.

No, it doesn't.

 I assumed the reader would understand it could be *any* kind of file. The only thing I found out is required is to have a parser initialized via the `vim.treesitter.language.add` for that particular file type.

I  used Go just as an example, since that was the kind of file I was using at the moment, nothing special.

Maybe there could be an even more minimal to reproduce this.

> 
> (It's really difficult to handle two people reporting slightly different issues at the same time. I appreciate you trying to help, but right now we're talking about _this_ issue, with a related by different example.)

I know, and I apologize if I contributed to the "noise".

I think @freddiehaddad created this new issue after commenting in the [previous one](https://github.com/neovim/neovim/issues/24186) and realizing no one was paying attention to it.

**Comment by clason** [comment]
> I assumed the reader would understand it could be any kind of file. The only thing I found out is required is to have a parser initialized via the vim.treesitter.language.add for that particular file type.

Another interesting point, but adds to the confusion. I'm not sure how that particular function could affect that.

A simple `vim.treesitter.start()` doesn't work?

**Comment by freddiehaddad** [comment]
> That is not minimal.

Okay, it doesn't get more minimal than these two files: https://github.com/neovim/neovim/issues/25608#issuecomment-1759725209

**Comment by freddiehaddad** [comment]
Sorry, just doublechecked and those examples do NOT work. I'll try to provide you a more minimal config. I'm assuming you don't want any plugin manger either?

**Comment by lucario387** [comment]
[Minimal-reproduction-template](https://github.com/neovim/neovim/wiki/Minimal-reproduction-template)

**Comment by itaranto** [comment]
> And you can reproduce the hang with just these two files and any version tracked Lua file? Because I cannot.

I think the key is to have a buffer opened for a file you are modifying from the ouside, for example, with Git.



> > I assumed the reader would understand it could be any kind of file. The only thing I found out is required is to have a parser initialized via the vim.treesitter.language.add for that particular file type.
> 
> Another interesting point, but adds to the confusion. I'm not sure how that particular function could affect that.
> 
> A simple `vim.treesitter.start()` doesn't work?

So far I cannot go more minimal than this:

```lua
vim.opt.foldmethod = "expr"
vim.opt.foldexpr = "v:lua.vim.treesitter.foldexpr()"

vim.treesitter.language.add("go", { path = "/home/ignacio/.local/share/nvim/lazy/nvim-treesitter/parser/go.so" })

vim.api.nvim_create_autocmd("FileType", {
	pattern = "*.go",
	callback = function(args)
		vim.treesitter.start(args.buf, "go")
	end,
})
```

(Replace the language to any language you want to test, remember to also modify the `pattern`  accordingly).

Also, to reproduce this, you don't actually need to use Git.

Assuming you are using `"go"` files, the repro steps are:
- `nvim --clean -u minimal.lua`
- Open `:e foo.go`
- In another terminal `echo hello >> foo.go`

Result:
Neovim hangs

**Comment by clason** [comment]
Thank you. I cannot reproduce the issue with that (on master), though.

**Comment by freddiehaddad** [comment]
Okay, this minimal config produces the issue:

`init.lua`

```lua
for name, url in pairs{
  treesitter = 'https://github.com/nvim-treesitter/nvim-treesitter'
} do
  local install_path = vim.fn.fnamemodify('nvim_issue/'..name, ':p')
  if vim.fn.isdirectory(install_path) == 0 then
    vim.fn.system { 'git', 'clone', '--depth=1', url, install_path }
  end
  vim.opt.runtimepath:append(install_path)
end

vim.opt.foldenable = false
vim.opt.foldmethod = 'expr'
-- https://github.com/neovim/neovim/issues/25608
vim.opt.foldexpr = 'v:lua.vim.treesitter.foldexpr()'
-- vim.opt.foldexpr = 'nvim_treesitter#foldexpr()'

require('nvim-treesitter.configs').setup({
    auto_install = true,
    highlight = {
        enable = true,
    },
    incremental_selection = {
        enable = true,
        keymaps = {
            init_selection = '<c-i>',
            node_incremental = '<c-i>',
            scope_incremental = false,
            node_decremental = '<bs>',
        },
    },
    indent = {
        enable = true,
    },
})
```

I will provide additional examples manipulating different settings.

**Comment by clason** [comment]
> I'll try a build with all bundled dependencies (which Arch's team removed recently) and I'll also try with the AppImage.

Yes, please try the appimage.  We cannot support distro builds, they make too many changes outside our control.

**Comment by freddiehaddad** [comment]
> Just to confirm: it is necessary to have nvim-treesitter installed, with everything enabled? Disabling even one thing will not let you reproduce?

Yes, it looks like treesitter must be installed. Don't think I've reproduced without it.  Still working through all my test scenarios to confirm.

In this example, I've disabled highlight, incremental_selection, and indent.  This issue still happens:

```lua
for name, url in pairs{
  treesitter = 'https://github.com/nvim-treesitter/nvim-treesitter'
} do
  local install_path = vim.fn.fnamemodify('nvim_issue/'..name, ':p')
  if vim.fn.isdirectory(install_path) == 0 then
    vim.fn.system { 'git', 'clone', '--depth=1', url, install_path }
  end
  vim.opt.runtimepath:append(install_path)
end

vim.opt.foldenable = false
vim.opt.foldmethod = 'expr'
-- https://github.com/neovim/neovim/issues/25608
vim.opt.foldexpr = 'v:lua.vim.treesitter.foldexpr()'
-- vim.opt.foldexpr = 'nvim_treesitter#foldexpr()'

require('nvim-treesitter.configs').setup({
    auto_install = true,
    highlight = {
        enable = false,
    },
    incremental_selection = {
        enable = false,
        keymaps = {
            init_selection = '<c-i>',
            node_incremental = '<c-i>',
            scope_incremental = false,
            node_decremental = '<bs>',
        },
    },
    indent = {
        enable = false,
    },
})
```

**Comment by clason** [comment]
Please test with the (nightly) appimage, not the Arch package.

**Comment by itaranto** [comment]
> Please test with the (nightly) appimage, not the Arch package.

The Nightly AppImage (`v0.10.0-dev-d2d3885`) doesn't have the issue while the `0.9.4` AppImage does.

**Comment by freddiehaddad** [comment]
> Can you try the minimal.lua file I provided?

Confirmed in a Go and Lua project and I can reproduce.

`minimal.lua`

```lua
vim.opt.foldmethod = "expr"
vim.opt.foldexpr = "v:lua.vim.treesitter.foldexpr()"

vim.treesitter.language.add("go", { path = "/home/freddiehaddad/.local/share/nvim/lazy/nvim-treesitter/parser/go.so" })

vim.api.nvim_create_autocmd("FileType", {
    pattern = "*.go",
    callback = function(args)
        vim.treesitter.start(args.buf, "go")
    end,
})
```

**Comment by clason** [comment]
> > Please test with the (nightly) appimage, not the Arch package.
> 
> The nightly AppImage (`v0.10.0-dev-d2d3885`) doesn't have the issue while the `0.9.4` AppImage does.

Thank you. Also found https://github.com/LazyVim/LazyVim/issues/1581, which corroborates that https://github.com/neovim/neovim/pull/24230 would have to be backported to fix this.

**Comment by tomtomjhj** [comment]
#24230 does not fix this issue. I have tried cherry-picking that commit on top of 0.9.4, but nvim still hangs. The logic "to avoid infinite recursion" was already there before that PR, and I added that comment only to make it clear.

**Comment by clason** [comment]
Is there a prior commit we forgot (or decided not) to backport?

**Comment by freddiehaddad** [comment]
@clason I learned a lot from this bug, thanks for your patience :)

- minimal config
- git bisect
- neovim treesitter coe vs nvim-treesetter plugin

**Comment by itaranto** [comment]
I was about to do a "reverse" bisect between `master` (at `d2d38858d1f33a2cd92e9a6182ece2459b0d6f75`) and `v0.9.0` which is the common ancestor between the `release-0.9.0` and the `master` branches, when I discovered that `v0.9.0` doesn't have the issue.

So, I did a "normal" bisect between `v0.9.0` and `v0.9.4 and I found the offending commit: `1b09bcef522754452e46294a2340050f390a590b`.

**Comment by konradmalik** [comment]
I used to have this issue and now it is resolved for me on the current master.

There is another similar bug though that I've found only when using [fugitive.vim](https://github.com/tpope/vim-fugitive) and its `:Git log` command. I was not able to reproduce it anywhere else. Also, Neovim does not hang indefinitely, but just takes a very long time to execute that command.

- with `vim.opt.foldexpr = "v:lua.vim.treesitter.foldexpr()"` - it hangs for 1-2 minutes before opening the log
- with `vim.opt.foldexpr = "nvim_treesitter#foldexpr()"` (or `vim.opt.foldmethod = indent`) - it opens the log immediately
